### ISSUE TITLE
LIBXSMM Mem Leak

### ISSUE THREAD

**Issue by jeremylt** [CEED/libCEED#212]
Linking against LIBXSMM is giving a small mem leak. This persists across any backend. To reproduce:

Build the library with only the LIBXSMM optional backend
`OCCA_DIR= MAGMA_DIR= CUDA_DIR= make AVX=0 prove`

Run any test in Valgrind:
`valgrind ./build/t000-init /cpu/self/ref/serial` (Occurs with any backend)

Output:
```
==5136== Memcheck, a memory error detector
==5136== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==5136== Using Valgrind-3.14.0.GIT and LibVEX; rerun with -h for copyright info
==5136== Command: ./build/t000-init /cpu/self/ref/serial
==5136== 
==5136== 
==5136== HEAP SUMMARY:
==5136==     in use at exit: 8 bytes in 1 blocks
==5136==   total heap usage: 8 allocs, 7 frees, 21,086,991 bytes allocated
==5136== 
==5136== LEAK SUMMARY:
==5136==    definitely lost: 0 bytes in 0 blocks
==5136==    indirectly lost: 0 bytes in 0 blocks
==5136==      possibly lost: 0 bytes in 0 blocks
==5136==    still reachable: 8 bytes in 1 blocks
==5136==         suppressed: 0 bytes in 0 blocks
==5136== Rerun with --leak-check=full to see details of leaked memory
==5136== 
==5136== For counts of detected and suppressed errors, rerun with: -v
==5136== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by jedbrown** [comment]
Please add `--leak-check=full --show-leak-kinds=all` (with a debug build) to see the line where that memory is allocated.

**Comment by jeremylt** [comment]
```
==10677== 8 bytes in 1 blocks are still reachable in loss record 1 of 1
==10677==    at 0x48376FF: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==10677==    by 0x60A898D: gomp_malloc (alloc.c:37)
==10677==    by 0x60B7C6B: gomp_init_num_threads (proc.c:91)
==10677==    by 0x60A6F0A: initialize_env (env.c:1244)
==10677==    by 0x4010549: call_init.part.0 (in /usr/lib/ld-2.28.so)
==10677==    by 0x4010649: _dl_init (in /usr/lib/ld-2.28.so)
==10677==    by 0x4002039: ??? (in /usr/lib/ld-2.28.so)
==10677==    by 0x1: ???
==10677==    by 0x1FFF000A42: ???
==10677==    by 0x1FFF000A54: ???
```

Still working on chasing the exact line number, but only happens when we link to LIBXSMM

**Comment by alheinecke** [comment]
I know @hfp is currently fixing some memory leaks in the backend of LIBXSMM. Perhaps this is related.

**Comment by hfp** [comment]
I will check and try to reproduce the issue.

**Comment by hfp** [comment]
I am not claiming to have reproduced the issue reported here (since I do not know which version of LIBXSMM is used), I only looked at hfp/master revision. What happens is:

```text
[...]
==279885== 32 bytes in 1 blocks are still reachable in loss record 1 of 1
==279885==    at 0x4C2B955: calloc (vg_replace_malloc.c:711)
==279885==    by 0x5D6561F: _dlerror_run (in /usr/lib64/libdl-2.17.so)
==279885==    by 0x5D65127: dlsym (in /usr/lib64/libdl-2.17.so)
[...]
```

It seems to be a know bug (cross-DSO cleanup issue with GLIBC/LIBDL/PTHREADS). I experimented a bit if there is a workaround triggering the cleanup somehow. I also looked at valgrind markers to suppress this issue. It might be that a newer GLIBC just resolves it. For an older GLIBC there seems to be no harm as this is supposed to be a pseudo-leak aka non-growing memory leak.
